examples/simplex.dx

```diff
@@ -0,0 +1,596 @@
+  (==) = \r1 r2.
+    case r1 of
+      {|variables=r1|} ->
+        case r2 of
+          {|variables=r2|} -> r1 == r2
+          {|slacks=r2|} -> False
+          {|value=r2|} -> False
+      {|slacks=r1|} ->
+        case r2 of
+          {|variables=r2|} -> False
+          {|slacks=r2|} -> r1 == r2
+          {|value=r2|} -> False
+      {|value=r1|} ->
+        case r2 of
+          {|variables=r2|} -> False
+          {|slacks=r2|} -> False
+          {|value=r2|} -> True
+
+
+instance [Eq a, Eq b] Eq {artificials: a | realvars: b}
+  (==) = \r1 r2.
+    case r1 of
+      {|artificials=r1|} ->
+        case r2 of
+          {|artificials=r2|} -> r1 == r2
+          {|realvars=r2|} -> False
+      {|realvars=r1|} ->
+        case r2 of
+          {|artificials=r2|} -> False
+          {|realvars=r2|} -> r1 == r2
+
+
+def argmax_suchthat [Ord o] (xs:n=>o) (cond:o->Bool): n =
+  -- Assumes that there's at least one element of xs
+  -- that satisfies the condition.
+  cmp = \x y. (not $ cond y) || (x > y && cond x)
+  argscan cmp xs
+
+def argmax_suchthat_table [Ord o] (xs:n=>o) (cond:n=>Bool): n =
+  -- This variant takes a table of pre-evaluated conditions.
+  cmp = \i j. (not $ cond.j) || (xs.i > xs.j && cond.i)
+  argscan cmp (for i. i)
+
+
+'### Simplex data types
+
+data SimplexResult n:Type =
+  Infeasible
+  Unbounded
+  Solution (n=>Float)
+
+data ConstraintType =
+  LessThan
+  GreaterThan
+  -- Todo: Add equality constraints
+
+data ObjectiveType =
+  Maximize
+  Minimize
+
+def Constraint (n:Type) : Type =
+  {coeffs:n=>Float & offset:Float & conType:ConstraintType}
+
+def ColIx (cons:Type) (vars:Type) : Type =
+  {variables:vars | slacks:cons | value: Unit}
+
+def RowIx (cons:Type) (a:Type) : Type =
+  {constraints:cons | objective:Unit | extra:a}
+
+Void : Type = Fin 0
+
+def AbsTableau (cons:Type) (vars:Type) (extra:Type) : Type =
+  coltype = ColIx cons vars
+  rowtype = RowIx cons extra
+  tabletype = rowtype => coltype => Float
+  (cons => coltype & vars => coltype & tabletype)
+  -- basics, nonbasics, table
+
+def Tableau (cons:Type) (vars:Type) : Type =
+  AbsTableau cons vars Void -- no need for extra row
+
+def FeasibilityTableau (cons:Type) (vars:Type) : Type =
+  -- The feasibility tableau has extra columns for the artificial variables.
+  -- Todo: use a List of artificial variables, since not each constraint
+  -- actually requires an artificial variable.
+  allvars = {realvars: vars | artificials: cons}
+  AbsTableau cons allvars Unit
+
+
+'### Simplex helper functions
+
+def eval_objective (coeffs: n=>Float) (x: n=>Float) : Float = dot coeffs x
+
+def constraints_satisfied (constraints: m=>Constraint n) (x: n=>Float) : Bool =
+  constraints_ok = all for c.
+    ({offset=offset, coeffs=coeffs, conType=conType}) = constraints.c
+    case conType of
+      LessThan    -> dot coeffs x <= offset
+      GreaterThan -> dot coeffs x >= offset
+  constraints_ok && all for i:n. x.i >= 0.0 || x.i ~~ 0.0
```

Comment:
Are you implicitly adding positivity constraints for all variables?

Reply:
Yep you are right! That is one assumption of this "basic" implementation. There was a note about this in the main `simplex` function, but I guess that's a long shot for any reader. I've added a few comments for readability.